### PR TITLE
BV : Run the minion stages in parallel

### PR DESCRIPTION
**Pull request by maximenoel8**
**WARNING**

This BV has been tested on local Jenkins using AWS provider.
It has not been tested yet on CI using libvirt provider.

**What this PR does ?**

This pull request modifies the behavior of the build validation pipeline.
The minions testing will now be conducted independently and concurrently.
In the event of a failed stage in one of the minions, its testing will be interrupted, but it will not impact the testing of the other minions.
To achieve this, the proxy setup was also refactored.
new pipeline

**Links:**
Issue : https://github.com/SUSE/spacewalk/issues/11337
Uyuni PR : https://github.com/uyuni-project/uyuni/pull/6479

**What's new ?**

- New resource : `mgrCreateBootstrapRepo` has been created. 

It is only possible to execute one instance of the `mgr create boostrap repo` command at a time. To prevent multiple minions from simultaneously calling this command on the server, this resource creates a queue.

- New rake call : create a rake run set for each minion that regroup the different stages: 
  - add maintenance update repositories
  - create bootstrap repository
  - add activation key
  - init client
  - smoke tests

maintenance update
add activation key
create bootstrap repo
init client
smoke tests

- Refactor proxy setup / testing : 
As the setup stages are now invoked separately by each minion, the proxy required a redesign to call these stages and prepare the environment. The newly created proxy stage is named Prepare Proxy and encompasses this sub-stages :
  - add maintenance update repositories
  - create bootstrap repository
  - add activation key

- Develop a function that aggregates all the stages of the minions and executes them concurrently

I decided to create a distinct function that groups all the stages of the minions and executes them in parallel. I believe this separation of parallel stages from sequential stages provides greater clarity and makes easier to identify the dynamic parallel stages and distinguish them from the sequential ones.

**How clientTestingStages works ?**

The first challenge is to obtain the list of minions at the Jenkins Worker level. My proposed solution is to utilize the `terraform state list` command to achieve this. Another option would be to connect to the controller and retrieve the bashrc file.

I have two lists of minions, one is the `nodeList` used by the testsuite, and the other is the `envVar` list used to configure the controller environment variables.

The stages of each minion will be stored in a HashMap named `tests`.

When calling rake cucumber, I use the command unset to open an ssh session only with none minion instances ( server, proxy, ... ) and the minion currently tested. `unset ${temporaryList.join(' ')};` will unset all the other minions.

The initial part of the function involves creating a `test` object for each minion. This object will store the sequential stages, configured to run only for that specific minion. Once all the minions have their individual `test` objects set up (`nodeList.each { minion -> tests["${minion}"] = {`), I invoke this collection using the parallel function. This will concurrently execute all the `tests` objects.

The `mgr create bootstrap repo` command is protected by the lock function (which must be available on the Jenkins CI), allowing only one instance of the command to run at a time. This step is executed within a try-catch block, releasing the resource if the creation of the bootstrap repository fails.

### CODE REVIEW

File: jenkins_pipelines/environments/common/pipeline-build-validation.groovy [hunk 1]
```diff
@@ -229,4 +202,160 @@ def run(params) {
...
+                        if(params.confirm_before_continue) {
+                            input 'Press any key to start creating bootstrap repositories'
+                        }
+                        // Employ a lock resource to prevent concurrent calls to create the bootstrap repository in the manager.
+                        // Utilize a try-catch mechanism to release the resource for other nodes in the event of a failed bootstrap.
+                        lock(resource: mgrCreateBootstrapRepo, timeout: 320) {
+                            try {
+                                echo 'Create bootstrap repository'
+                                res_create_bootstrap_repository = sh(script: "./terracumber-cli ${common_params} --logfile ${resultdirbuild}/testsuite.log --runstep cucumber --cucumber-cmd 'unset ${temporaryList.join(' ')}; export BUILD_VALIDATION=true; cd /root/spacewalk/testsuite; rake cucumber:build_validation_create_bootstrap_repository_${minion}'", returnStatus: true)
+                                echo "Create bootstrap repository code: ${res_create_bootstrap_repository}"
```
Comment (nodeg) [inline jenkins_pipelines/environments/common/pipeline-build-validation.groovy:312]:
```suggestion
                                echo "Create bootstrap repository status code: ${res_create_bootstrap_repository}"
```

File: jenkins_pipelines/environments/common/pipeline-build-validation.groovy [hunk 2]
```diff
@@ -229,4 +202,160 @@ def run(params) {
...
+                                if (res_create_bootstrap_repository != 0) {
+                                    error("Create bootstrap repository fail: ${res_add_keys}")
```
Comment (nodeg) [inline jenkins_pipelines/environments/common/pipeline-build-validation.groovy:314]:
```suggestion
                                    error("Create bootstrap repository failed with status code: ${res_create_bootstrap_repository}")
```

**Review by nodeg — CHANGES_REQUESTED** [review]
Thank you very much this big improvement Maxime! Nice work. I left some comments/nitpicks/suggestions.

File: jenkins_pipelines/environments/common/pipeline-build-validation.groovy [hunk 3]
```diff
@@ -42,6 +45,8 @@ def run(params) {
                     sh "set +x; source /home/jenkins/.credentials set -x; export TF_VAR_CUCUMBER_GITREPO=${params.cucumber_gitrepo}; export TF_VAR_CUCUMBER_BRANCH=${params.cucumber_ref}; export TERRAFORM=${params.terraform_bin}; export TERRAFORM_PLUGINS=${params.terraform_bin_plugins}; ./terracumber-cli ${common_params} --logfile ${resultdirbuild}/sumaform.log ${env.TERRAFORM_INIT} --taint '.*(domain|main_disk|data_disk|server_extra_nfs_mounts).*' --custom-repositories ${WORKSPACE}/custom_repositories.json --sumaform-backend ${params.sumaform_backend} --runstep provision"
                     // Generate features
                     sh "./terracumber-cli ${common_params} --logfile ${resultdirbuild}/testsuite.log --runstep cucumber --cucumber-cmd 'export BUILD_VALIDATION=true; cd /root/spacewalk/testsuite; rake utils:generate_build_validation_features'"
+                    // Generate rake files
+                    sh "./terracumber-cli ${common_params} --logfile ${resultdirbuild}/testsuite.log --runstep cucumber --cucumber-cmd 'export BUILD_VALIDATION=true; cd /root/spacewalk/testsuite; rake jenkins:generate_rake_files_build_validation'"
```
Comment (srbarrios) [inline jenkins_pipelines/environments/common/pipeline-build-validation.groovy:49]: I would safely remove "rake_namespace" variable from the pipeline parameters.
And replace wherever we use it by cucumber.
As I would say it doesn't make sense any more to use this other kind of parallelization via gem.
And we can just use parallelization via Jenkins pipelines?

Make sense for you?

File: jenkins_pipelines/environments/common/pipeline-build-validation.groovy [hunk 4]
```diff
@@ -229,4 +202,160 @@ def run(params) {
     }
```
Comment (srbarrios) [inline jenkins_pipelines/environments/common/pipeline-build-validation.groovy:202]: I'm curious, the merge of reports looks good I guess?

**Review by srbarrios — APPROVED** [review]
Just a remark of not using double parallelization by using not only jenkins parallel but also the Ruby gem parallel_tests.

Aside that, I'm curious to see how it looks regarding the report.

But approving already.

File: jenkins_pipelines/environments/common/pipeline-build-validation.groovy (hunk 3 above)
Comment (maximenoel8) [reply jenkins_pipelines/environments/common/pipeline-build-validation.groovy:49]: Yes indeed, I was thinking about it yesterday.

**Comment by maximenoel8** [comment]
Report example :

[cucumber_report.zip](https://github.com/SUSE/susemanager-ci/files/10738431/cucumber_report.zip)

**Review by nodeg — APPROVED** [review]

**Comment by srbarrios** [comment]
🤘🥳🤘

**Pull request merged**